Added new effects

Jaydawg8888 opened the pull request:
Pinkerton Protection:
- Creates a total of 20 pinkerton npc's that'll orbit around the player.
- 3 seconds for an entire rotation to happen.
- Effect lasts 60 seconds.

Baby Horse:
- All horses near the player will be shrunk to a scale of 0.5 from 1.0, they'll automatically be re-scaled upon effect disablement. Note: If the horse leaves the player's vicinity they'll stay shrunk until the effect has concluded. 
- Effect lasts 45 seconds.

Can't Tie Shoes:
- If the player is on foot and they've been running for over 2 seconds, they'll be thrown into a ragdoll state. 2 second grace is given in-case they accidently tap sprint or whatever!
- Effect lasts 30 seconds.

No Minimap:
- Disables the Radar & Hud overall for 60 seconds, will automatically be shown again after the effect has ended.
- Effect lasts 60 seconds.

Anonymous Benefactor:
- Clears any active bounty & pursuit on the player.
- Instant (not timed).

Let's Take A Look:
- Forces the Binoculars to be equipped into the player's hand, if they don't have the binoculars then it'll be given to them.

Teleport To Tumbleweed:
- Teleports the player (and any mount they're on) to Tumebleweed.
- Instant (not timed).

Teleport To Closest Train Station:
- Teleports the player to the nearest train station (this includes post offices and non post office locations).
- Instant (not timed).

Teleport To Tallest Mountain (Country Roads Take Me Home)
- Teleports the player to the tallest mountain that is traversable back to ground level. 
- Instant (not timed).

Clean All Weapons:
- All weapons currently on the player's person are set to clean.
- Instant (not timed).

Dirty All Weapons:
- All weapons currently on the player's person are set to fully dirty.
- Instant (not timed).

Freeze For 5 Seconds (e.g. Slow World For 5 Seconds)
- Game speed is set to a crawl (minimum that is possible), all peds are frozen and invincible during this time.
- Timed event for 5 seconds.

Set All Player Horse Stats (Max, Half & Min):
- All horse stats such as Bondness, Stamina & Health are set to whatever the effect is.
- Note: Sometimes the stats exceed level 10, I assume this is because of other factors such as saddle etc), I've only tested it with the beginner horse, haven't tried any other others, but the effect does seem fine?
Any feedback on this would be good!
- Instant (not timed).

Horse Magnet
- All horses rather mounted by a ped or not (except Wagons) will follow the player until the timed event is over, after that their task will reset and go back to doing whatever it was they were doing.
- Timed event for 60 seconds.

Gunlinger Duel:
- Spawns an enemy gunslinder with a Cattleman Revolver, the NPC is super accurate but does 50% less damage than normal (for more fairness). 
- Instant (not timed).

Fighting Ghosts:
- Spawns 5 invisible NPC's (Aberdeen Pig Farmer), they're invisible and have 500 HP, you can outrun them quite easily.
- If the player moves out of the vicinity of the ghosts, they'll be deleted and more will be spawned, once the timer is up they'll all be deleted.
- Effect lasts 30 seconds.

Old West Noir:
- Plays the Noir effect from the Photo Mode camera. Takes approx 2 seconds for the effect to fully come into effect, we lerp the strength over 2 seconds so it's not just an instant boom to your face!
- Effect lasts 90 seconds.

Horse Loses All Stamina (Can't Teach A Horse To Drink):
- The horse that the player is mounted on (if not then the horse that they have active) will have it's stamina drained.
- Instant (not timed).

Tree Hugger:
- Spawns a tree in-front of the player, if they're on a mount they're dismounted immediately,  plays a hugging animation on loop until the effect has ended.
- Effect lasts 10 seconds.

Inverted Gravity:
- Updated Inverted Gravity to no longer kill the player when they take falldamage!

Other:
- Updated gitignore to exclude .idea generated files from JetBrains Rider

Jaydawg8888 commented:
@archanyhm  @chrisschwartzdev  Added you both as reviewers - I'm not going to be pushing anymore effects for now as I've got to do some mod work for DarkViper on the GTA V side.

Let me know if there's any issues or changes you'd like made <3

chrisschwartzdev commented on src/Effects/peds.cpp:
```diff
@@ -2113,3 +2119,367 @@ void EffectSpawnExtremeEvilMicah::OnTick()
+
+	Vector3 playerLocation = ENTITY::GET_ENTITY_COORDS(PLAYER::PLAYER_PED_ID(), true, 0);
+
+	std::set<Ped> toDelete;
+	
+	for (auto ghost : ghostPeds)
+	{
+		if (!ENTITY::DOES_ENTITY_EXIST(ghost) || PED::IS_PED_DEAD_OR_DYING(ghost, false))
+		{
+			// ghost doesn't exist or is dead - so make a new one
+			toDelete.insert(ghost);
+			
+			SpawnGhost();
+			continue;
+		}
+
+		float distance = GetDistance3D(playerLocation, ENTITY::GET_ENTITY_COORDS(ghost, true, 0));
+
+		if (distance <= maxDistanceFromPlayer)
+		{
+			// player is still within range of the ghost
+			continue;
+		}
+
+		// player is too far away from the ghost, so delete it and spawn a new one
+		toDelete.insert(ghost);
+		SpawnGhost();
+	}
+
+	for (auto ghost : toDelete)
+	{
+		if (ENTITY::DOES_ENTITY_EXIST(ghost))
+		{
+			PED::DELETE_PED(&ghost);
+		}
+		
+		ghostPeds.erase(ghost);
+	}
```
This is okay how it is, but I don't think it's necessary to have two loops here if you use an iterator instead - you would be able to erase dead / out of range peds while iterating `ghostPeds` only once.

chrisschwartzdev commented on src/Effects/peds.h:
```diff
@@ -793,6 +793,160 @@ class EffectSpawnExtremeEvilMicah : public Effect
+	virtual void OnActivate() override;
+
+private:
+	inline static std::vector<const char*> gunSlinderModels = {"CS_famousgunslinger_01", "CS_famousgunslinger_02"
```
Typo: `gunSlinderModels`

chrisschwartzdev reviewed (approved):
Approved - although I suggest removing "No Minimap" since it's a duplicate of "No HUD".

--------------------

Pinkerton Protection:
* Looks good, the pinkertons also shield the player as expected

Baby Horse:
* Looks good but it crashed the script the first time I enabled it - might be my PC though, maybe a good idea to test again or have someone else test..

Can't Tie Shoes:
* Works well
* Something funny happened after the effect expired - I thought sprinting wasn't working anymore, but in fact, John had broken his leg xD

No Minimap:
* 🚩 Appears to hide more than the minimap - in fact it seems like it hides the entire HUD. I see in the code, you noted this fact, but this effect is no different than the existing "No HUD" effect.

Anonymous Benefactor:
* Works only after a crime is reported, but doesn't clear witnesses - which I think is expected?

Let's Take A Look:
* Works

Teleport To Tumbleweed:
* Works

Teleport To Closest Train Station:
* Works

Teleport To Tallest Mountain (Country Roads Take Me Home)
* Works

Clean All Weapons:
* Works

Dirty All Weapons:
* Works

Freeze For 5 Seconds (e.g. Slow World For 5 Seconds)
* Seems to work as expected

Set All Player Horse Stats (Max, Half & Min):
* Seems to work fine (all 3 effects)

Horse Magnet
* Works well

Gunlinger Duel:
* Works

Fighting Ghosts:
* Works well (and is quite funny lol)

Old West Noir:
* Works well

Horse Loses All Stamina (Can't Teach A Horse To Drink):
* Works

Tree Hugger:
* Works great

Inverted Gravity:
* Works - it did kill my stamina, but did not kill me, so I think that's fine.

Updated gitignore to exclude .idea generated files from JetBrains Rider
* Fellow Rider user! Just a heads up, you can ignore these locally by editing your `.git/info/exclude` file, instead of pushing it to the repo, if you want to.

Jaydawg8888 commented:
> Approved - although I suggest removing "No Minimap" since it's a duplicate of "No HUD".

Good spot didn't notice that there was already a duplicate effect - I've removed the "No Minimap" effect from the list.

`Typo: gunSlinderModels`
My quick typing finally caught up to me ): Fixed this ;p

`This is okay how it is, but I don't think it's necessary to have two loops here if you use an iterator instead - you would be able to erase dead / out of range peds while iterating ghostPeds only once.`

I've updated this to now be a single iteration, similar to how you've done your Rabbit code ;p I don't touch C++ too much so this has been a good learning curve.

> Updated gitignore to exclude .idea generated files from JetBrains Rider
Fellow Rider user! Just a heads up, you can ignore these locally by editing your .git/info/exclude file, instead of pushing it to the repo, if you want to.

Good to know for the future! Might as well keep it in as there's no harm in it, but I shall look into the above for future stuff.

Now I need to figure out how I'm going to fix merge conflicts

chrisschwartzdev commented:
@Jaydawg8888 

Yeah no worries, feel free to keep the `.idea` change.

As far as merging, honestly it might be a good idea to squash everything in your PR to one commit and then rebase and deal with the merge conflicts, then you'll only have to deal with it once instead of 20 times.

Jaydawg8888 commented:
@chrisschwartzdev  Tried to squash using SourceTree, idk if it actually worked as expected lol.

I did resolve the Merge Conflicts however, I think it looks all good? If you have some spare time at all please do give it a one over just to make sure I didn't do anything wrong. (I don't really do squashing/conflict resolving often )-;

chrisschwartzdev commented:
Yeah it doesn't look like you did it quite right. I'll explain how I do it, but before you do any of this, create a new branch from this one so you don't lose anything if it goes wrong `git checkout -b backup`.

This is how I squash commits, I do it almost every day at my job. Just follow these steps exactly.

1. You're going to use the command `git reset HEAD~<n>`, where <n> is the number of commits on your branch (currently 34). What this does is basically reset your HEAD to -<n> commits, but keeping all of the changes from those commits. So the final command would be `git reset HEAD~34`.
2. Now you want to make your commit to squash them. So `git commit` and then you can add the details for all of the effects you added.
3. Now comes the scary part, but if you backed up your branch, you shouldn't have anything to worry about. Your local branch and the remote branch will be out of sync by <n> commits, because locally you basically overwrote all of those commits with one. After you commit it will say something like `Your branch and ... have diverged, and have <x> and <y> different commits each, respectively.`. At this point you want to force push to update the remote with your local branch with `git push -f`.
4. If you haven't rebased onto upstream yet (the `master` branch of this repo), you'd do that now, since you only have one commit to worry about merge conflicts. So you can do a `git merge upstream/master` or a `git rebase upstream/master` (assuming that's how you have your remotes set up) and resolve the conflicts now, if any.
5. Finally, if you rebased, you'll see a message like `Your branch and ... have diverged` again, but if you merged and do `git status`, you'll see another commit added that's just a merge commit. At this point you can force push again with `git push -f` if you rebased, or just `git push` if you merged.
6. From github, now you should be able to merge the PR without any conflicts.

It's up to you whether you rebase or merge - rebase will end up looking cleaner, but merge might end up being a bit easier. Hope that helps, and let me know if you run into any issues.

chrisschwartzdev commented:
Or actually thinking about it again, if you already resolve the conflicts, then you could probably just merge from github. There might be an extra commit or two in there, but as far as I can see, all of your effects are there. I would make sure it builds before you do merge it though.

Jaydawg8888 commented:
@chrisschwartzdev I've managed to get a version of what you mentioned above here:
https://github.com/Jaydawg8888/Chaos-Mod-RDR2---DarkViperAU-Version/tree/backup2

How would I go about rebasing that onto master? Or what would you suggest ;p - First time doing it through git commands as I normally use SourceTree so learning new stuff as I go!

chrisschwartzdev commented:
@Jaydawg8888 It looks like your branch `backup2` is already up to date - maybe what I said was confusing, but the squashing should have happened on this branch (master) instead of the backup branch, the backup was just in case commits got lost on master.

Anyways, assuming you have no other work on master (basically I assume that master is _behind_ backup2 now), you should be able to switch to master (`git checkout master`) then do `git reset backup2 --hard` and then `git push -f`. What this will do is make your master branch have the same commits as backup2. That should update the PR and you should then be able to merge this PR on github.

Jaydawg8888 commented:
@chrisschwartzdev  Ahhh, I understand now - My mistake, I misunderstood the instructions.

Either way, I have ran the commands above and it has been updated.

Thank you again so much for your help

chrisschwartzdev commented:
@Jaydawg8888 Yeah it seems good to merge the PR now. Glad I could help!
